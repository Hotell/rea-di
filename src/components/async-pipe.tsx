import { Component, PureComponent, ReactNode } from 'react'

type Props<T> = {
  value: Promise<T>
  children(api: ReturnType<AsyncPipe<T>['getApi']>): ReactNode
}
type State<T> = Readonly<{
  resolvedValue: T
  isLoading: boolean
}>
export function asyncPipe<T>(value: Promise<T>, componentInstance: Component) {
  return value.then((resolved) => {
    componentInstance.forceUpdate(() => {
      console.log('force update')
    })

    return resolved
  })
}

export class AsyncPipe<T> extends PureComponent<Props<T>, State<T>> {
  readonly state = {
    isLoading: true,
    resolvedValue: (null as any) as T,
  }
  private getApi() {
    const { isLoading, resolvedValue: resolved } = this.state

    return {
      resolved,
      isLoading,
    }
  }
  render() {
    // console.log('render with', this.getApi())
    const { children } = this.props

    return children(this.getApi())
  }
  componentDidMount() {
    this.resolvePromise()
  }
  // componentDidUpdate(prevProps: AsyncPipeProps<T>) {
  //   const promiseChanged = prevProps.value !== this.props.value
  //   console.log('componentDidUpdate changed?', promiseChanged)
  //   if (promiseChanged) {
  //     this.setState(() => ({ isLoading: true }), () => this.resolvePromise())
  //   }
  // }
  private resolvePromise() {
    this.props.value
      .then((resolved) => {
        this.setState((prevState) => ({
          isLoading: false,
          resolvedValue: resolved,
        }))
      })
      .catch(console.error)
  }
}
