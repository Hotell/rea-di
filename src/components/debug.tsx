import {
  Provider as ProviderConfig,
  ReflectiveInjector,
  Type,
} from 'injection-js'
import React, { ReactElement, ReactNode } from 'react'

import { rootInjector } from '../services/injector-context'
import { isProvider, isType } from '../utils/guards'

export const Debug = (props: {
  enable: boolean
  parentInjector: ReflectiveInjector
  children: ReactNode
  registeredProviders: ProviderConfig[]
  label?: string
}) => {
  const { children, enable, label, registeredProviders, parentInjector } = props

  if (!enable) {
    return children as ReactElement<any>
  }

  const isRoot = parentInjector === rootInjector
  const injectorLabel = label || isRoot ? 'Root Injector' : 'Child Injector'
  const bgColor = isRoot ? 'red' : '#388e3c'
  const styling = {
    container: { border: `2px solid ${bgColor}`, padding: '.5rem' },
    header: { backgroundColor: bgColor, padding: `.5rem .25rem` },
    title: { margin: 0, backgroundColor: bgColor },
  }

  const registeredProvidersNames: string[] = getRegisteredProvidersNames(
    registeredProviders
  )

  return (
    <div style={styling.container}>
      <header style={styling.header}>
        <h4 style={styling.title}>{injectorLabel}</h4>
        <pre>
          <b>Registered Providers:</b> {json(registeredProvidersNames)}
        </pre>
      </header>
      {children}
    </div>
  )

  function json<T>(value: T) {
    // tslint:disable-next-line:no-magic-numbers
    return JSON.stringify(value, null, 2)
  }

  function getRegisteredProvidersNames(providers: ProviderConfig[]): string[] {
    return providers.reduce((acc, next) => {
      if (isType(next)) {
        return [...(acc as string[]), next.name]
      }

      if (isProvider(next)) {
        const [registrationKey] = Object.keys(next).filter(
          (val) => val !== 'provide'
        )

        const registrationValue = (next as {
          [key: string]: any
        })[registrationKey] as Type<any> | object

        const providerName = {
          provide: next.provide.name ? next.provide.name : next.provide._desc,
          as: isType(registrationValue)
            ? registrationValue.name
            : JSON.stringify(registrationValue),
        }

        return [
          ...(acc as string[]),
          `{provide: ${providerName.provide}, ${registrationKey}: ${
            providerName.as
          } }`,
        ]
      }

      return acc
    }, []) as string[]
  }
}
