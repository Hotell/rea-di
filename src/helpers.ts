import { ComponentType } from 'react'

export const getComponentDisplayName = <P>(Component: ComponentType<P>) =>
  Component.displayName ||
  Component.name ||
  (Component.constructor && Component.constructor.name) ||
  'Component'
