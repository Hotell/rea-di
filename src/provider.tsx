import {
  Provider as ProviderConfig,
  ReflectiveInjector,
  Type,
} from 'injection-js'
import React, { PureComponent, ReactElement, ReactNode } from 'react'

import { Context, ContextApi, rootInjector } from './context'
import { isProvider, isType } from './guards'
import { StateCallback } from './types'

const Debug = (props: {
  parentInjector: ReflectiveInjector
  children: ReactNode
  registeredProviders: ProviderConfig[]
  label?: string
}) => {
  const { children, label, registeredProviders, parentInjector } = props

  // tslint:disable-next-line:no-use-before-declare
  if (!Provider._debugMode.on) {
    return children as ReactElement<any>
  }

  const isRoot = parentInjector === rootInjector
  const injectorLabel = label || isRoot ? 'Root Injector' : 'Child Injector'
  const bgColor = isRoot ? 'red' : '#388e3c'
  const styling = {
    container: { border: `2px solid ${bgColor}`, padding: '.5rem' },
    header: { backgroundColor: bgColor, padding: `.5rem .25rem` },
    title: { margin: 0, backgroundColor: bgColor },
  }

  const registeredProvidersNames: string[] = getRegisteredProvidersNames(
    registeredProviders
  )

  return (
    <div style={styling.container}>
      <header style={styling.header}>
        <h4 style={styling.title}>{injectorLabel}</h4>
        <pre>
          <b>Registered Providers:</b> {json(registeredProvidersNames)}
        </pre>
      </header>
      {children}
    </div>
  )

  function json<T>(value: T) {
    // tslint:disable-next-line:no-magic-numbers
    return JSON.stringify(value, null, 2)
  }

  function getRegisteredProvidersNames(providers: ProviderConfig[]): string[] {
    return providers.reduce((acc, next) => {
      if (isType(next)) {
        return [...(acc as string[]), next.name]
      }

      if (isProvider(next)) {
        const [registrationKey] = Object.keys(next).filter(
          (val) => val !== 'provide'
        )

        const registrationValue = (next as {
          [key: string]: any
        })[registrationKey] as Type<any> | object

        const providerName = {
          provide: next.provide.name ? next.provide.name : next.provide._desc,
          as: isType(registrationValue)
            ? registrationValue.name
            : JSON.stringify(registrationValue),
        }

        return [
          ...(acc as string[]),
          `{provide: ${providerName.provide}, ${registrationKey}: ${
            providerName.as
          } }`,
        ]
      }

      return acc
    }, []) as string[]
  }
}

type Props = {
  children: ReactElement<any>
  provide: ProviderConfig[]
}
type State = ContextApi

/**
 *
 */
export class Provider extends PureComponent<Props, State> {
  static _debugMode = {
    on: false,
  }
  static enableDebugMode() {
    this._debugMode.on = true
  }
  private static Debug = Debug

  private injector?: ReflectiveInjector
  readonly state: State = {} as State
  private providersRegistered = false

  private monkeyPatchStateProviders(
    injector: ReflectiveInjector,
    providers: ProviderConfig[]
  ) {
    type TypeWithState = Type<any> & { setState(...args: any[]): any }

    providers.reduce((acc, next) => {
      let stateKey: string
      let provideClass!: TypeWithState
      if (isType(next)) {
        stateKey = next.name
        provideClass = next as TypeWithState
      }
      if (isProvider(next)) {
        const { provide } = next
        stateKey = provide.name
        provideClass = provide
      }

      const hasState =
        isType(provideClass) && 'setState' in provideClass.prototype

      if (hasState) {
        // console.log('HAS STATE')
        const instance = injector.get(provideClass)
        const originalSetStateFn = instance.setState.bind(instance)

        const newSetStateFn = (state: StateCallback) => {
          // call service setState
          const newState = originalSetStateFn(state) as object

          // update context provider state
          this.setState(
            (prevState) => {
              return { [stateKey]: newState }
            }
            // () => originalSetStateFn(state)
          )
        }

        instance.setState = newSetStateFn
      }

      return acc
    }, {})
  }

  private get contextApi(): ContextApi {
    return {
      injector: this.injector,
      ...this.state,
    }
  }
  private renderProvider = ({ injector: parentInjector }: ContextApi) => {
    this.injector =
      this.injector || parentInjector.resolveAndCreateChild(this.props.provide)

    if (!this.providersRegistered) {
      this.monkeyPatchStateProviders(this.injector, this.props.provide)
    }

    this.providersRegistered = true

    if (Provider._debugMode.on) {
      return (
        <Context.Provider value={this.contextApi}>
          <Provider.Debug
            parentInjector={parentInjector}
            registeredProviders={this.props.provide}
          >
            {this.props.children}
          </Provider.Debug>
        </Context.Provider>
      )
    }

    return (
      <Context.Provider value={this.contextApi}>
        {this.props.children}
      </Context.Provider>
    )
  }

  render() {
    return <Context.Consumer>{this.renderProvider}</Context.Consumer>
  }
}
