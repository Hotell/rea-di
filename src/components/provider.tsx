import {
  Provider as ProviderConfig,
  ReflectiveInjector,
  Type,
} from 'injection-js'
import React, { PureComponent, ReactElement } from 'react'

import { IS_PROD } from '../environment'
import { Context, ContextApi } from '../services/injector-context'
import { StateCallback } from '../types'
import { isProvider, isType } from '../utils/guards'
import { Debug } from './debug'

type Props = {
  children: ReactElement<any>
  providers: ProviderConfig[]
}
type State = ContextApi

/**
 *
 */
export class DependencyProvider extends PureComponent<Props, State> {
  private static _debugMode = {
    on: false,
  }
  static enableDebugMode() {
    if (IS_PROD) {
      // tslint:disable-next-line:no-console
      console.info('ðŸ‘‰ DEBUG MODE IS AVAILABLE ONLY IN NON PROD ENV')

      return
    }

    this._debugMode.on = true
  }

  private injector?: ReflectiveInjector
  readonly state: State = {} as State
  private providersRegistered = false

  private monkeyPatchStateProviders(
    injector: ReflectiveInjector,
    providers: ProviderConfig[]
  ) {
    type TypeWithState = Type<any> & { setState(...args: any[]): any }

    providers.reduce((acc, next) => {
      let stateKey: string
      let provideClass!: TypeWithState
      if (isType(next)) {
        stateKey = next.name
        provideClass = next as TypeWithState
      }
      if (isProvider(next)) {
        const { provide } = next
        stateKey = provide.name
        provideClass = provide
      }

      const hasState =
        isType(provideClass) && 'setState' in provideClass.prototype

      if (hasState) {
        const instance = injector.get(provideClass)
        const originalSetStateFn = instance.setState.bind(instance)

        const newSetStateFn = (state: StateCallback) => {
          // call service setState
          const newState = originalSetStateFn(state) as object

          // update context provider state
          this.setState(
            (prevState) => {
              return { [stateKey]: newState }
            }
            // () => originalSetStateFn(state)
          )
        }

        instance.setState = newSetStateFn
      }

      return acc
    }, {})
  }

  private get contextApi(): ContextApi {
    return {
      injector: this.injector,
      ...this.state,
    }
  }
  private renderProvider = ({ injector: parentInjector }: ContextApi) => {
    const isInDebugMode = !IS_PROD && DependencyProvider._debugMode.on

    this.injector =
      this.injector ||
      parentInjector.resolveAndCreateChild(this.props.providers)

    if (!this.providersRegistered) {
      this.monkeyPatchStateProviders(this.injector, this.props.providers)
    }

    this.providersRegistered = true

    if (isInDebugMode) {
      return (
        <Context.Provider value={this.contextApi}>
          <Debug
            enable
            parentInjector={parentInjector}
            registeredProviders={this.props.providers}
          >
            {this.props.children}
          </Debug>
        </Context.Provider>
      )
    }

    return (
      <Context.Provider value={this.contextApi}>
        {this.props.children}
      </Context.Provider>
    )
  }

  render() {
    return <Context.Consumer>{this.renderProvider}</Context.Consumer>
  }
}
