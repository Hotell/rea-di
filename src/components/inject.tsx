import React, { Component, ReactNode } from 'react'

import { reflection } from '../facade/lang'
import { Context, ContextApi } from '../services/injector-context'
import { Constructor, NullableInstanceTypes } from '../types'
import { metadataKey } from '../utils/helpers'

type InjectProps<C extends Array<Constructor | null>> = {
  values: C
  children: (...injectables: NullableInstanceTypes<C>) => ReactNode
}

export class Inject<T extends Array<Constructor | null>> extends Component<
  InjectProps<T>
> {
  private injectMappedProviders = ({ injector }: ContextApi) => {
    const injectables = this.props.values.map((nextInjectableRef) => {
      if (!nextInjectableRef) {
        return null
      }

      const annotationsMeta = reflection.getMetadata(
        metadataKey,
        nextInjectableRef
      ) as undefined | { optional: boolean }

      if (!annotationsMeta) {
        return injector.get(nextInjectableRef)
      }

      const isOptional = annotationsMeta && annotationsMeta.optional
      const notFoundValue = isOptional ? null : undefined
      const wrappedInjectableRef = ((nextInjectableRef as any) as () => Constructor)()

      return injector.get(wrappedInjectableRef, notFoundValue)
    }) as NullableInstanceTypes<T>

    return this.props.children(...(injectables as any))
  }
  render() {
    return <Context.Consumer>{this.injectMappedProviders}</Context.Consumer>
  }
}
