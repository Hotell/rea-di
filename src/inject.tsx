import { getMetadata } from '@abraham/reflection'
import React, { Component, ReactNode } from 'react'

import { Context, ContextApi } from './context'
import { metadataKey } from './helpers'
import { Constructor, NullableInstanceTypes } from './types'

type InjectProps<C extends Array<Constructor | null>> = {
  values: C
  children: (...injectables: NullableInstanceTypes<C>) => ReactNode
}

export class Inject<T extends Array<Constructor | null>> extends Component<
  InjectProps<T>
> {
  private injectMappedProviders = ({ injector }: ContextApi) => {
    const injectables = this.props.values.map((nextInjectableRef) => {
      if (!nextInjectableRef) {
        return null
      }

      const annotationsMeta = getMetadata(metadataKey, nextInjectableRef) as
        | undefined
        | { optional: boolean }

      if (!annotationsMeta) {
        return injector.get(nextInjectableRef)
      }

      const isOptional = annotationsMeta && annotationsMeta.optional
      const notFoundValue = isOptional ? null : undefined
      const wrappedInjectableRef = ((nextInjectableRef as any) as () => Constructor)()

      return injector.get(wrappedInjectableRef, notFoundValue)
    }) as NullableInstanceTypes<T>

    return this.props.children(...(injectables as any))
  }
  render() {
    return <Context.Consumer>{this.injectMappedProviders}</Context.Consumer>
  }
}
