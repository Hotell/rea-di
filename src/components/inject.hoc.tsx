import React, { Component, ComponentType } from 'react'

import { Type } from 'injection-js'
import { HoC, NullableInstanceTypes, NullableTypeMap, Subtract } from '../types'
import { createHOCName } from '../utils/helpers'
import { Inject } from './inject'

/**
 * If you need to access injected service instances outside of render, you can use this high order component.
 * It will give you also a little performance boost because providers map is gonna be created only once, during definition.
 * This can be mitigated by extracting providers to instance property when standard <Inject values={tuple(...)}/> is used.
 *
 * @param tokenMap - provider instances from injector to be mapped to wrapped component props
 */
export const withInjectables = <TokenMap extends NullableTypeMap>(
  tokenMap: TokenMap
) => <
  OriginalProps extends NullableInstanceTypes<TokenMap>,
  ResolvedProps = Subtract<OriginalProps, TokenMap>
>(
  Cmp: ComponentType<OriginalProps>
): HoC<ResolvedProps, typeof Cmp> => {
  const injectablePropsKeys: Array<keyof TokenMap> = Object.keys(tokenMap)
  const tokenRefs = injectablePropsKeys.map((propKey) => tokenMap[propKey])

  class WithInjectables extends Component<ResolvedProps> {
    static displayName: string = createHOCName(WithInjectables, Cmp)
    static readonly WrappedComponent = Cmp

    /**
     *
     * @param providersMap - map config (key to Class type)
     * @param propsKeys - array of mapped keys
     * @param injectables - injectables are nullable tuples
     */
    private injectTokensMap(
      providersMap: TokenMap,
      propsKeys: Array<keyof TokenMap>,
      injectables: NullableInstanceTypes<Array<Type<any>>>
    ) {
      const injectProps = propsKeys.reduce((acc, nextPropKey) => {
        // token is always present. We just trick TS to always be an Nullable type to get proper inference within children
        const token = providersMap[nextPropKey] as Type<any>
        const injectable =
          injectables.find(
            (injectableInstance) => injectableInstance instanceof token
          ) || null

        return { ...acc, [nextPropKey]: injectable }
      }, {}) as NullableInstanceTypes<TokenMap>

      return injectProps
    }

    render() {
      const { ...rest } = this.props as object /* FIXED in ts 3.2 */

      return (
        <Inject values={tokenRefs}>
          {(...injectables) => (
            <Cmp
              {...this.injectTokensMap(
                tokenMap,
                injectablePropsKeys,
                injectables
              )}
              {...rest}
            />
          )}
        </Inject>
      )
    }
  }

  return WithInjectables
}
