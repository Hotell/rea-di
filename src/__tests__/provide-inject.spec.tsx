// tslint:disable:no-magic-numbers
// tslint:disable:jsx-no-lambda
// tslint:disable:no-shadowed-variable
import React, { Component } from 'react'

import { mount } from 'enzyme'
import { noop } from '../helpers'
import { Inject } from '../inject'
import { Provider } from '../provider'
import { Counter } from './setup/components'
import { CounterService, Logger } from './setup/services'
import { select } from './setup/utils'

class CounterModule extends Component<{ title: string }> {
  render() {
    return (
      <div>
        <h3>{this.props.title}</h3>
        <Inject providers={{ logger: Logger, counterService: CounterService }}>
          {(injectables) => {
            return (
              <Counter
                counterService={injectables.counterService}
                logger={injectables.logger}
              >
                Hello projection
              </Counter>
            )
          }}
        </Inject>
      </div>
    )
  }
}

const App = () => {
  return (
    <main>
      <Provider provide={[Logger, CounterService]}>
        <CounterModule title="count module" />
      </Provider>
    </main>
  )
}

describe(`Provide/Inject`, () => {
  it(`should properly inject`, () => {
    const tree = mount(<App />)

    const counter = tree.find(Counter)
    const counterChildren = counter.prop('children')

    expect(counterChildren!.toString()).toBe('Hello projection')

    expect(tree).toBeTruthy()
    expect(tree).toMatchSnapshot()

    tree.unmount()
  })

  it(`should properly update state on stateful service and reflect it on component tree`, () => {
    const tree = mount(<App />)

    const counter = tree.find(Counter)
    const counterLogger = counter.props().logger
    const incButton = counter.find('button').at(0)
    const decButton = counter.find('button').at(1)
    const countParagraph = counter.find('p')

    jest.spyOn(counterLogger, 'log').mockImplementation(noop)

    expect(countParagraph.text()).toBe('Count: 0')
    expect(counterLogger.log).not.toHaveBeenCalled()

    incButton.simulate('click')

    expect(countParagraph.text()).toBe('Count: 1')
    expect(counterLogger.log).toHaveBeenCalledTimes(1)

    decButton.simulate('click')
    decButton.simulate('click')

    expect(countParagraph.text()).toBe('Count: -1')
    expect(counterLogger.log).toHaveBeenCalledTimes(3)

    tree.unmount()
  })

  it(`should properly create hierarchical injectors`, () => {
    class LoggerMock implements Logger {
      log = jest.fn((...args: any[]) => void 0)
    }
    // tslint:disable:prefer-const
    let parentLoggerInstance!: LoggerMock
    let childLoggerInstance!: LoggerMock

    const App = () => {
      return (
        <Provider provide={[{ provide: Logger, useClass: LoggerMock }]}>
          <div data-test="parentInjector">
            <Inject providers={{ logger: Logger }}>
              {(fromParent) => {
                parentLoggerInstance = fromParent.logger as LoggerMock

                return (
                  <div>
                    <button
                      onClick={() => {
                        fromParent.logger.log('from parent')
                      }}
                    >
                      log from parent
                    </button>
                    <Provider
                      provide={[{ provide: Logger, useClass: LoggerMock }]}
                    >
                      <div data-test="childInjector">
                        <Inject providers={{ logger: Logger }}>
                          {(fromChild) => {
                            childLoggerInstance = fromChild.logger as LoggerMock

                            return (
                              <div>
                                <button
                                  onClick={() => {
                                    fromParent.logger.log('parent from child')
                                  }}
                                >
                                  log parent from child
                                </button>
                                <button
                                  onClick={() => {
                                    fromChild.logger.log('from child')
                                  }}
                                >
                                  log from child
                                </button>
                              </div>
                            )
                          }}
                        </Inject>
                      </div>
                    </Provider>
                  </div>
                )
              }}
            </Inject>
          </div>
        </Provider>
      )
    }

    const tree = mount(<App />)

    const parentBtn = tree.find(`${select('parentInjector')} button`).at(0)
    const childrenButtons = tree.find(`${select('childInjector')} button`)
    const childrenBtnThatLogsParent = childrenButtons.at(0)
    const childrenBtnThatLogsChild = childrenButtons.at(1)

    expect(parentLoggerInstance.log).not.toHaveBeenCalled()

    parentBtn.simulate('click')

    expect(parentLoggerInstance.log).toHaveBeenCalledWith('from parent')

    childrenBtnThatLogsParent.simulate('click')

    expect(parentLoggerInstance.log).toHaveBeenLastCalledWith(
      'parent from child'
    )
    expect(parentLoggerInstance.log).toHaveBeenCalledTimes(2)

    expect(childLoggerInstance.log).not.toHaveBeenCalled()

    childrenBtnThatLogsChild.simulate('click')

    expect(parentLoggerInstance.log).toHaveBeenCalledTimes(2)
    expect(childLoggerInstance.log).toHaveBeenCalledTimes(1)
    expect(childLoggerInstance.log).toHaveBeenLastCalledWith('from child')
  })
})
