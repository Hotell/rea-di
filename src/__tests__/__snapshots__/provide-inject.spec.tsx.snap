// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Provide/Inject should properly inject 1`] = `
<App>
  <main>
    <Provider
      provide={
        Array [
          [Function],
          [Function],
        ]
      }
    >
      <CounterModule
        title="count module"
      >
        <div>
          <h3>
            count module
          </h3>
          <Inject
            providers={
              Object {
                "counterService": [Function],
                "logger": [Function],
              }
            }
          >
            <Counter
              counterService={
                CounterService {
                  "logger": Logger {},
                  "setState": [Function],
                  "state": Object {
                    "count": 0,
                  },
                }
              }
              logger={Logger {}}
            >
              <div
                className="counter"
              >
                <h4>
                  Counter
                </h4>
                <p>
                  Count: 
                  0
                </p>
                <button
                  onClick={[Function]}
                >
                  increment
                </button>
                <button
                  onClick={[Function]}
                >
                  decrement
                </button>
                <section
                  className="children-section"
                >
                  Hello projection
                </section>
              </div>
            </Counter>
          </Inject>
        </div>
      </CounterModule>
    </Provider>
  </main>
</App>
`;
